xbmc/threads/test/TestEndTime.cpp

```diff
@@ -10,41 +10,43 @@
 
 #include <gtest/gtest.h>
 
+using namespace std::chrono_literals;
+
 namespace
 {
 
-void CommonTests(XbmcThreads::EndTime& endTime)
+void CommonTests(XbmcThreads::EndTime<>& endTime)
 {
-  EXPECT_EQ(static_cast<unsigned int>(100), endTime.GetInitialTimeoutValue());
-  EXPECT_LT(static_cast<unsigned int>(0), endTime.GetStartTime());
+  EXPECT_EQ(100ms, endTime.GetInitialTimeoutValue());
+  EXPECT_LT(0ms, endTime.GetStartTime());
 
   EXPECT_FALSE(endTime.IsTimePast());
-  EXPECT_LT(static_cast<unsigned int>(0), endTime.MillisLeft());
+  EXPECT_LT(0ms, endTime.GetTimeLeft());
 
-  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  std::this_thread::sleep_for(101ms);
```

Comment:
100ms, not 101 !

Reply:
This was deliberate as the test doesn't pass otherwise

```
$ ./kodi-test --gtest_filter="TestEndTime*" 
Note: Google Test filter = TestEndTime*
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from TestEndTime
[ RUN      ] TestEndTime.DefaultConstructor
[       OK ] TestEndTime.DefaultConstructor (100 ms)
[ RUN      ] TestEndTime.ExplicitConstructor
/home/lukas/Documents/git/xbmc/xbmc/threads/test/TestEndTime.cpp:28: Failure
Value of: endTime.IsTimePast()
  Actual: false
Expected: true
[  FAILED  ] TestEndTime.ExplicitConstructor (101 ms)
[----------] 2 tests from TestEndTime (201 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (250 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TestEndTime.ExplicitConstructor

 1 FAILED TEST
```

Reply:
```
    if (now <= (m_startTime + m_totalWaitTime))
      return false;
 ```
 Do you mean btw. "<" - that's why the 100ms test does not work above I assume. If you start at 0 and duration is 100 ms, then when the clock shows 100, you still return false, which is not correct.

Reply:
That's a good observation but I believe it has the desired result. If the duration is `100ms` and you wait `100ms` that shouldn't be considered past, you need to wait `101ms` to be considered past.

I believe the difference comes from using the `std::chrono::time_point_cast<T>`. This changes the `time_point` to be in the templated type. The default `time_point` is using `std::chrono::nanoseconds`. This makes it so that the actual checked value is truncated to the units of measurement.

This is demonstrated in this sample programs. https://godbolt.org/z/xYq6nf3PE

with `using unit = std::chrono::nanoseconds;`
```
future: 30344022422788ns
now: 30344022500756ns
30344022500756ns > 30344022422788ns
time_waited: 100077968ns
100077968ns > 100ms
```

with `using unit = std::chrono::milliseconds;`
```
future: 30672482ms
now: 30672482ms
30672482ms == 30672482ms
time_waited: 100ms
100ms == 100ms
```

So I guess the discussion should be that if it's desired to wait `100ms` should it be considered with the smallest time point possible or with the unit desired to wait? I'll probably need to think about this for a bit to come to a conclusion however I am thinking that any unit of time passed the desired wait time should be considered past in which case the current behaviour in this PR is wrong. I'm unsure how this was working before as everything was in `milliseconds` with an `int`.

Reply:
Before this PR:
```
return ((std::chrono::steady_clock::now() - m_startTime) >= m_totalWaitTime);
```

Old: now - start >= duration
New: now <= (start + duration) <=> now - start <= duration for negative case => now - start > duration for the positive case. So, yes behaviour changed from >= to >.

That's why I posted the comment above.

Reply:
And to add: to get it "exact" in ms range - according to Shannon Nicquist, clock precission at least double as exact as ms is needed :-) ...

Reply:
> Before this PR:
> 
> ```
> 
> return ((std::chrono::steady_clock::now() - m_startTime) >= m_totalWaitTime);
> 
> ```
> 
> 
> 
> Old: now - start >= duration
> 
> New: now <= (start + duration) <=> now - start <= duration for negative case => now - start > duration for the positive case. So, yes behaviour changed from >= to >.
> 
> 
> 
> That's why I posted the comment above. 

I'm sorry I don't follow. The sample program shows the same behavior with both methods and only that the precision of the clock influences the behavior. What am I missing?

Reply:
From reading the code, the new code changed the behaviour for the "==" case. Old code: IsTimePast() returned true for this case, new code returns false. Or did I read this wrong?

Reply:
> From reading the code, the new code changed the behaviour for the "==" case. Old code: IsTimePast() returned true for this case, new code returns false. Or did I read this wrong?

Yes you are correct.